### ISSUE TITLE
[1.21, 1.21.1] Forced disconnection on a NeoForge dedicated server when looking at water (and potentially any fluid block)

### ISSUE THREAD

**Issue by 3TUSK** [McJty/TheOneProbe#655]
# Synopsis 

The One Probe can cause disconnection on dedicated server when looking at water (and potentially any fluid block) under certain conditions.

# How-to reproduce

1. Setup a dedicated server using 
   - Minecraft 1.21
   - NeoForge 21.1.68
   - The One Probe 1.21_neo-12.0.3
2. Prepare a client instance using 
   - Minecraft 1.21
   - NeoForge 21.1.68
   - The One Probe 1.21_neo-12.0.3
3. On client, edit `theoneprobe-client.toml`: set `showLiquids` to `true`.
4. Start the server
5. Launch the game client
6. Join the server
7. Find a water block
8. Look at it

# Expected behavior

![image](https://github.com/user-attachments/assets/3b6c59e0-ffc2-4154-a6d4-1d6e7c91d760)

# Actual behavior

Client was forced to quit the server; server log contains the following error:

<details>
<summary>Server error & stacktrace</summary>

```
[12:06:26] [Netty Server IO #1/ERROR] [minecraft/Connection]: Exception caught in connection
io.netty.handler.codec.EncoderException: Failed to encode packet 'clientbound/minecraft:custom_payload'
	at TRANSFORMER/minecraft@1.21.1/net.minecraft.network.codec.IdDispatchCodec.encode(IdDispatchCodec.java:55) ~[server-1.21.1-20240808.144430-srg.jar%23120!/:?] {re:classloading}
	at TRANSFORMER/minecraft@1.21.1/net.minecraft.network.codec.IdDispatchCodec.encode(IdDispatchCodec.java:14) ~[server-1.21.1-20240808.144430-srg.jar%23120!/:?] {re:classloading}
	at TRANSFORMER/neoforge@21.1.68/net.neoforged.neoforge.network.filters.GenericPacketSplitter.encode(GenericPacketSplitter.java:104) ~[neoforge-21.1.68-universal.jar%23121!/:?] {re:classloading}
	at TRANSFORMER/neoforge@21.1.68/net.neoforged.neoforge.network.filters.GenericPacketSplitter.encode(GenericPacketSplitter.java:39) ~[neoforge-21.1.68-universal.jar%23121!/:?] {re:classloading}
	at MC-BOOTSTRAP/io.netty.codec@4.1.97.Final/io.netty.handler.codec.MessageToMessageEncoder.write(MessageToMessageEncoder.java:90) ~[netty-codec-4.1.97.Final.jar%23104!/:4.1.97.Final] {}
	at MC-BOOTSTRAP/io.netty.transport@4.1.97.Final/io.netty.channel.AbstractChannelHandlerContext.invokeWrite0(AbstractChannelHandlerContext.java:881) ~[netty-transport-4.1.97.Final.jar%23108!/:4.1.97.Final] {}
	at MC-BOOTSTRAP/io.netty.transport@4.1.97.Final/io.netty.channel.AbstractChannelHandlerContext.invokeWrite(AbstractChannelHandlerContext.java:863) ~[netty-transport-4.1.97.Final.jar%23108!/:4.1.97.Final] {}
	at MC-BOOTSTRAP/io.netty.transport@4.1.97.Final/io.netty.channel.AbstractChannelHandlerContext.write(AbstractChannelHandlerContext.java:968) ~[netty-transport-4.1.97.Final.jar%23108!/:4.1.97.Final] {}
	at MC-BOOTSTRAP/io.netty.transport@4.1.97.Final/io.netty.channel.AbstractChannelHandlerContext.write(AbstractChannelHandlerContext.java:856) ~[netty-transport-4.1.97.Final.jar%23108!/:4.1.97.Final] {}
	at MC-BOOTSTRAP/io.netty.codec@4.1.97.Final/io.netty.handler.codec.MessageToMessageEncoder.write(MessageToMessageEncoder.java:113) ~[netty-codec-4.1.97.Final.jar%23104!/:4.1.97.Final] {}
	at MC-BOOTSTRAP/io.netty.transport@4.1.97.Final/io.netty.channel.AbstractChannelHandlerContext.invokeWrite0(AbstractChannelHandlerContext.java:881) ~[netty-transport-4.1.97.Final.jar%23108!/:4.1.97.Final] {}
	at MC-BOOTSTRAP/io.netty.transport@4.1.97.Final/io.netty.channel.AbstractChannelHandlerContext.invokeWrite(AbstractChannelHandlerContext.java:863) ~[netty-transport-4.1.97.Final.jar%23108!/:4.1.97.Final] {}
	at MC-BOOTSTRAP/io.netty.transport@4.1.97.Final/io.netty.channel.AbstractChannelHandlerContext.write(AbstractChannelHandlerContext.java:968) ~[netty-transport-4.1.97.Final.jar%23108!/:4.1.97.Final] {}
	at MC-BOOTSTRAP/io.netty.transport@4.1.97.Final/io.netty.channel.AbstractChannelHandlerContext.write(AbstractChannelHandlerContext.java:856) ~[netty-transport-4.1.97.Final.jar%23108!/:4.1.97.Final] {}
	at MC-BOOTSTRAP/io.netty.transport@4.1.97.Final/io.netty.channel.ChannelOutboundHandlerAdapter.write(ChannelOutboundHandlerAdapter.java:113) ~[netty-transport-4.1.97.Final.jar%23108!/:4.1.97.Final] {}
	at TRANSFORMER/minecraft@1.21.1/net.minecraft.network.Connection$2.write(Connection.java:528) ~[server-1.21.1-20240808.144430-srg.jar%23120!/:?] {re:classloading}
	at MC-BOOTSTRAP/io.netty.transport@4.1.97.Final/io.netty.channel.AbstractChannelHandlerContext.invokeWrite0(AbstractChannelHandlerContext.java:881) ~[netty-transport-4.1.97.Final.jar%23108!/:4.1.97.Final] {}
	at MC-BOOTSTRAP/io.netty.transport@4.1.97.Final/io.netty.channel.AbstractChannelHandlerContext.invokeWriteAndFlush(AbstractChannelHandlerContext.java:940) ~[netty-transport-4.1.97.Final.jar%23108!/:4.1.97.Final] {}
	at MC-BOOTSTRAP/io.netty.transport@4.1.97.Final/io.netty.channel.AbstractChannelHandlerContext.write(AbstractChannelHandlerContext.java:966) ~[netty-transport-4.1.97.Final.jar%23108!/:4.1.97.Final] {}
	at MC-BOOTSTRAP/io.netty.transport@4.1.97.Final/io.netty.channel.AbstractChannelHandlerContext.writeAndFlush(AbstractChannelHandlerContext.java:934) ~[netty-transport-4.1.97.Final.jar%23108!/:4.1.97.Final] {}
	at MC-BOOTSTRAP/io.netty.transport@4.1.97.Final/io.netty.channel.AbstractChannelHandlerContext.writeAndFlush(AbstractChannelHandlerContext.java:984) ~[netty-transport-4.1.97.Final.jar%23108!/:4.1.97.Final] {}
	at MC-BOOTSTRAP/io.netty.transport@4.1.97.Final/io.netty.channel.DefaultChannelPipeline.writeAndFlush(DefaultChannelPipeline.java:1025) ~[netty-transport-4.1.97.Final.jar%23108!/:4.1.97.Final] {}
	at MC-BOOTSTRAP/io.netty.transport@4.1.97.Final/io.netty.channel.AbstractChannel.writeAndFlush(AbstractChannel.java:306) ~[netty-transport-4.1.97.Final.jar%23108!/:4.1.97.Final] {}
	at TRANSFORMER/minecraft@1.21.1/net.minecraft.network.Connection.doSendPacket(Connection.java:364) ~[server-1.21.1-20240808.144430-srg.jar%23120!/:?] {re:classloading}
	at TRANSFORMER/minecraft@1.21.1/net.minecraft.network.Connection.lambda$sendPacket$12(Connection.java:359) ~[server-1.21.1-20240808.144430-srg.jar%23120!/:?] {re:classloading}
	at MC-BOOTSTRAP/io.netty.common@4.1.97.Final/io.netty.util.concurrent.AbstractEventExecutor.runTask(AbstractEventExecutor.java:174) ~[netty-common-4.1.97.Final.jar%23105!/:4.1.97.Final] {}
	at MC-BOOTSTRAP/io.netty.common@4.1.97.Final/io.netty.util.concurrent.AbstractEventExecutor.safeExecute(AbstractEventExecutor.java:167) ~[netty-common-4.1.97.Final.jar%23105!/:4.1.97.Final] {}
	at MC-BOOTSTRAP/io.netty.common@4.1.97.Final/io.netty.util.concurrent.SingleThreadEventExecutor.runAllTasks(SingleThreadEventExecutor.java:470) ~[netty-common-4.1.97.Final.jar%23105!/:4.1.97.Final] {}
	at MC-BOOTSTRAP/io.netty.transport@4.1.97.Final/io.netty.channel.nio.NioEventLoop.run(NioEventLoop.java:569) ~[netty-transport-4.1.97.Final.jar%23108!/:4.1.97.Final] {}
	at MC-BOOTSTRAP/io.netty.common@4.1.97.Final/io.netty.util.concurrent.SingleThreadEventExecutor$4.run(SingleThreadEventExecutor.java:997) ~[netty-common-4.1.97.Final.jar%23105!/:4.1.97.Final] {}
	at MC-BOOTSTRAP/io.netty.common@4.1.97.Final/io.netty.util.internal.ThreadExecutorMap$2.run(ThreadExecutorMap.java:74) ~[netty-common-4.1.97.Final.jar%23105!/:4.1.97.Final] {}
	at java.base/java.lang.Thread.run(Thread.java:1583) [?:?] {}
Caused by: java.lang.RuntimeException: Failed encoding custom payload theoneprobe:returninfo: java.lang.NullPointerException: Cannot invoke "net.minecraft.resources.ResourceLocation.toString()" because "p_130086_" is null
	at TRANSFORMER/minecraft@1.21.1/net.minecraft.network.protocol.common.custom.CustomPacketPayload$1.writeCap(CustomPacketPayload.java:42) ~[server-1.21.1-20240808.144430-srg.jar%23120!/:?] {re:classloading}
	at TRANSFORMER/minecraft@1.21.1/net.minecraft.network.protocol.common.custom.CustomPacketPayload$1.encode(CustomPacketPayload.java:47) ~[server-1.21.1-20240808.144430-srg.jar%23120!/:?] {re:classloading}
	at TRANSFORMER/minecraft@1.21.1/net.minecraft.network.protocol.common.custom.CustomPacketPayload$1.encode(CustomPacketPayload.java:29) ~[server-1.21.1-20240808.144430-srg.jar%23120!/:?] {re:classloading}
	at TRANSFORMER/minecraft@1.21.1/net.minecraft.network.codec.StreamCodec$4.encode(StreamCodec.java:81) ~[server-1.21.1-20240808.144430-srg.jar%23120!/:?] {re:classloading}
	at TRANSFORMER/minecraft@1.21.1/net.minecraft.network.codec.StreamCodec$5.encode(StreamCodec.java:97) ~[server-1.21.1-20240808.144430-srg.jar%23120!/:?] {re:classloading}
	at TRANSFORMER/minecraft@1.21.1/net.minecraft.network.codec.StreamCodec$5.encode(StreamCodec.java:87) ~[server-1.21.1-20240808.144430-srg.jar%23120!/:?] {re:classloading}
	at TRANSFORMER/minecraft@1.21.1/net.minecraft.network.codec.IdDispatchCodec.encode(IdDispatchCodec.java:53) ~[server-1.21.1-20240808.144430-srg.jar%23120!/:?] {re:classloading}
	... 31 more
Caused by: java.lang.NullPointerException: Cannot invoke "net.minecraft.resources.ResourceLocation.toString()" because "p_130086_" is null
	at TRANSFORMER/minecraft@1.21.1/net.minecraft.network.FriendlyByteBuf.writeResourceLocation(FriendlyByteBuf.java:571) ~[server-1.21.1-20240808.144430-srg.jar%23120!/:?] {re:classloading}
	at TRANSFORMER/theoneprobe@1.21_neo-12.0.3/mcjty.theoneprobe.apiimpl.elements.ElementIcon.toBytes(ElementIcon.java:66) ~[theoneprobe-1.21_neo-12.0.3.jar%23122!/:1.21_neo-12.0.3] {re:classloading}
	at TRANSFORMER/theoneprobe@1.21_neo-12.0.3/mcjty.theoneprobe.apiimpl.ProbeInfo.writeElements(ProbeInfo.java:64) ~[theoneprobe-1.21_neo-12.0.3.jar%23122!/:1.21_neo-12.0.3] {re:classloading}
	at TRANSFORMER/theoneprobe@1.21_neo-12.0.3/mcjty.theoneprobe.apiimpl.elements.AbstractElementPanel.toBytes(AbstractElementPanel.java:74) ~[theoneprobe-1.21_neo-12.0.3.jar%23122!/:1.21_neo-12.0.3] {re:classloading}
	at TRANSFORMER/theoneprobe@1.21_neo-12.0.3/mcjty.theoneprobe.apiimpl.ProbeInfo.writeElements(ProbeInfo.java:64) ~[theoneprobe-1.21_neo-12.0.3.jar%23122!/:1.21_neo-12.0.3] {re:classloading}
	at TRANSFORMER/theoneprobe@1.21_neo-12.0.3/mcjty.theoneprobe.apiimpl.elements.AbstractElementPanel.toBytes(AbstractElementPanel.java:74) ~[theoneprobe-1.21_neo-12.0.3.jar%23122!/:1.21_neo-12.0.3] {re:classloading}
	at TRANSFORMER/theoneprobe@1.21_neo-12.0.3/mcjty.theoneprobe.apiimpl.ProbeInfo$1.encode(ProbeInfo.java:36) ~[theoneprobe-1.21_neo-12.0.3.jar%23122!/:1.21_neo-12.0.3] {re:classloading}
	at TRANSFORMER/theoneprobe@1.21_neo-12.0.3/mcjty.theoneprobe.apiimpl.ProbeInfo$1.encode(ProbeInfo.java:20) ~[theoneprobe-1.21_neo-12.0.3.jar%23122!/:1.21_neo-12.0.3] {re:classloading}
	at TRANSFORMER/minecraft@1.21.1/net.minecraft.network.codec.StreamCodec$9.encode(StreamCodec.java:181) ~[server-1.21.1-20240808.144430-srg.jar%23120!/:?] {re:classloading}
	at TRANSFORMER/minecraft@1.21.1/net.minecraft.network.protocol.common.custom.CustomPacketPayload$1.writeCap(CustomPacketPayload.java:40) ~[server-1.21.1-20240808.144430-srg.jar%23120!/:?] {re:classloading}
	at TRANSFORMER/minecraft@1.21.1/net.minecraft.network.protocol.common.custom.CustomPacketPayload$1.encode(CustomPacketPayload.java:47) ~[server-1.21.1-20240808.144430-srg.jar%23120!/:?] {re:classloading}
	at TRANSFORMER/minecraft@1.21.1/net.minecraft.network.protocol.common.custom.CustomPacketPayload$1.encode(CustomPacketPayload.java:29) ~[server-1.21.1-20240808.144430-srg.jar%23120!/:?] {re:classloading}
	at TRANSFORMER/minecraft@1.21.1/net.minecraft.network.codec.StreamCodec$4.encode(StreamCodec.java:81) ~[server-1.21.1-20240808.144430-srg.jar%23120!/:?] {re:classloading}
	at TRANSFORMER/minecraft@1.21.1/net.minecraft.network.codec.StreamCodec$5.encode(StreamCodec.java:97) ~[server-1.21.1-20240808.144430-srg.jar%23120!/:?] {re:classloading}
	at TRANSFORMER/minecraft@1.21.1/net.minecraft.network.codec.StreamCodec$5.encode(StreamCodec.java:87) ~[server-1.21.1-20240808.144430-srg.jar%23120!/:?] {re:classloading}
	at TRANSFORMER/minecraft@1.21.1/net.minecraft.network.codec.IdDispatchCodec.encode(IdDispatchCodec.java:53) ~[server-1.21.1-20240808.144430-srg.jar%23120!/:?] {re:classloading}
	... 31 more
```

</details>

# Analysis

https://github.com/McJtyMods/TheOneProbe/blob/2710f43dd4e670df85ef2a089ee532ef5fc6132e/src/main/java/mcjty/theoneprobe/apiimpl/providers/DefaultProbeInfoProvider.java#L349-L350

As a client extension, `IClientFluidTypeExtensions` are only registered on phyiscal client[^1]. When you call `IClientFluidTypeExtensions.of(fluid)` on dedicated server, you will end up with `IClientFluidTypeExtensions.DEFAULT`. 

Calling `getStillTexture` on `IClientFluidTypeExtensions.DEFAULT` gives you `null`[^2], leading to the NPE above and client disconnection. 
Similarly, this error is not reproducible neither in single-player world nor a LAN multi-player world because both use integrated server. You may have guessed that the screenshot in "Expected Behavior" section is taken in a single-player world. 

A proper fix requires retrieving this information on physical client instead, i.e. sending the fluid registry name instead, and off-loading the texture location query to physical client. 

[^1]: https://github.com/neoforged/NeoForge/blob/f869db584b46fd49066f266ee768fe51f071c836/src/main/java/net/neoforged/neoforge/client/ClientNeoForgeMod.java#L93-L161
[^2]: https://github.com/neoforged/NeoForge/blob/f869db584b46fd49066f266ee768fe51f071c836/src/main/java/net/neoforged/neoforge/client/extensions/common/IClientFluidTypeExtensions.java#L77

**Comment by McJty** [comment]
Thanks for the detailed report. I'll check it out soon

**Comment by McJty** [comment]
Fixed in 12.0.4
